contributors/design-proposals/api-paging.md

```diff
@@ -0,0 +1,148 @@
+* The `kube-apiserver` also has to deserialize that response into a single object, and then re-serialize it back to the client
+  * Much of the decoded etcd memory is copied into the struct used to serialize to the client
+* A client like `kubectl get` will then decode the response from JSON or protobuf
+  * A client with a slow connection may not be able to receive the entire response body within the default 60s timeout
+  * The recently introduced client compression feature can assist
+  * The large response will also be loaded entirely into memory
+
+The standard solution for reducing the impact of large reads is to allow them to be broken into smaller reads via a technique commonly referred to as paging. By efficiently splitting large list ranges from clients to etcd into many smaller list ranges, we can reduce the maximum memory allocation on etcd and the apiserver.
+
+Our primary consistent store etcd3 offers support for efficient paging with minimal overhead, and mechanisms exist for other potential future stores such as SQL databases or Consul to also implement a simple form of paging.
+
+Relevant issues:
+
+* https://github.com/kubernetes/kubernetes/issues/2349
+
+## Proposed change:
+
+Expose a simple paging mechanism to allow large API responses to be broken into consistent partial responses. Clients would indicate a tolerance for paging (opt-in) by specifying a desired maximum number of results to return in a `LIST` call. The server would return up to that amount of objects, and if more exist it would return a `continue` parameter that the client could pass to receive the next set of results.  The server would be allowed to ignore the limit if it does not implement limiting (backward compatible), but it is not allowed to support limiting without supporting a way to continue the query past the limit (may not implement `limit` without `continue`).
+
+```
+GET /api/v1/pods?limit=500
+{
+  "metadata": {"continue": "ABC...", "resourceVersion": "147"},
+  "items": [
+     // no more than 500 items
+   ]
+}
+GET /api/v1/pods?limit=500&continue=ABC...
+{
+  "metadata": {"continue": "DEF...", "resourceVersion": "147"},
+  "items": [
+     // no more than 500 items
+   ]
+}
+GET /api/v1/pods?limit=500&continue=DEF...
+{
+  "metadata": {"resourceVersion": "147"},
+  "items": [
+     // no more than 500 items
+   ]
+}
+```
+
+The token returned by the server for `continue` would be an opaque serialized string that would contain a simple serialization of the initially requested limit, a version identifier (to allow future extension), and any additional data needed by the server storage to identify where to start the next range. 
+
+The schema of the continue token is chosen by the storage layer and is not guaranteed to remain consistent for clients - clients *must* consider the continue token as opaque. Server implementations *should* ensure that continue tokens can persist across server restarts and across upgrades.
```

Comment:
Is there any guarantee w.r.t. state of the list?  For example, while we assume that the "continute" token tells us where keep going, is there any agreement that the rest of the list is fixed based on when the initial query was done or can it change?  This is an important detail since at some point we may add a 'size' field to the initial response and I don't think we want to change the semantics of the overall query in a future release.  Or, as another example, two queries at the same time but one client is slower to get all chunks - should the results look the same or will the slower one possible get different data because the system changed after the faster one finished.  We just need to be clear on what clients should see.